code/modules/mob/living/carbon/xenomorph/abilities/crusher/crusher_abilities.dm

```diff
@@ -62,3 +74,193 @@
+		INVOKE_ASYNC(X, /atom/movable.proc/throw_atom, lol, momentum/2, SPEED_FAST, null, TRUE)
+		stop_momentum()
+		return
+	if(!isturf(X.loc))
+		stop_momentum()
+		return
+	// Don't build up charge if you move via getting propelled by something
+	if(X.throwing)
+		stop_momentum()
+		return
+
+	var/do_stop_momentum = FALSE
+
+	// Need to be constantly moving in order to maintain charge
+	if(world.time > last_charge_move + 0.5 SECONDS)
+		do_stop_momentum = TRUE
+	if(dir != charge_dir)
+		charge_dir = dir
+		do_stop_momentum = TRUE
+
+	if(do_stop_momentum)
+		stop_momentum()
+	if(X.plasma_stored <= plasma_per_step)
+		stop_momentum()
+		return
+	last_charge_move = world.time
+	steps_taken++
+	if(steps_taken < steps_to_charge)
+		return
+	if(momentum < max_momentum)
+		momentum++
+		ADD_TRAIT(X, TRAIT_CHARGING, TRAIT_SOURCE_XENO_ACTION_CHARGE)
+		X.update_icons()
+		if(momentum == max_momentum)
+			X.emote("roar")
+	//X.use_plasma(plasma_per_step) // take if you are in toggle charge mode
+	if(momentum > 0)
+		X.use_plasma(plasma_per_step) // take plasma when you have momentum
+
+	noise_timer = noise_timer ? --noise_timer : 3
+	if(noise_timer == 3)
+		playsound(X, 'sound/effects/alien_footstep_charge1.ogg', 50)
+
+	for(var/mob/living/carbon/human/M in X.loc)
+		if(M.lying && M.stat != DEAD)
+			X.visible_message(SPAN_DANGER("[X] runs [M] over!"),
+				SPAN_DANGER("You run [M] over!")
+			)
+
+			M.apply_damage(momentum * 10)
+			animation_flash_color(M)
+
+	X.recalculate_speed()
+
+/datum/action/xeno_action/onclick/charger_charge/proc/handle_dir_change(datum/source, old_dir, new_dir)
+	SIGNAL_HANDLER
+	if(new_dir != charge_dir)
+		charge_dir = new_dir
+		if(momentum)
+			stop_momentum()
+
+/datum/action/xeno_action/onclick/charger_charge/proc/handle_river(datum/source, covered)
+	SIGNAL_HANDLER
+	if(!covered)
+		stop_momentum()
+
+/datum/action/xeno_action/onclick/charger_charge/proc/update_speed(mob/living/carbon/Xenomorph/X)
+	SIGNAL_HANDLER
+	X.speed += momentum * speed_per_momentum
+
+/datum/action/xeno_action/onclick/charger_charge/proc/stop_momentum(datum/source)
+	SIGNAL_HANDLER
+	var/mob/living/carbon/Xenomorph/X = owner
+	if(momentum == max_momentum)
+		X.visible_message(SPAN_DANGER("[X] skids to a halt!"))
+
+	REMOVE_TRAIT(X, TRAIT_CHARGING, TRAIT_SOURCE_XENO_ACTION_CHARGE)
+	steps_taken = 0
+	momentum = 0
+	X.recalculate_speed()
+	X.update_icons()
+
+/datum/action/xeno_action/onclick/charger_charge/proc/lose_momentum(amount)
+	if(amount >= momentum)
+		stop_momentum()
+	else
+		momentum -= amount
+		var/mob/living/carbon/Xenomorph/X = owner
+		X.recalculate_speed()
+
+/datum/action/xeno_action/onclick/charger_charge/proc/handle_collision(mob/living/carbon/Xenomorph/X, atom/A)
+	SIGNAL_HANDLER
+	if(!momentum)
+		stop_momentum()
+		return
+
+	var/result = A.handle_charge_collision(X, src)
+	switch(result)
+		if(XENO_CHARGE_TRY_MOVE)
+			if(step(X, charge_dir))
+				return COMPONENT_LIVING_COLLIDE_HANDLED
```

Comment:
More descriptive variable name here please. Is A the target that's being collided into? If it is, make it clear.
Same with X. Is that the charger or some different xeno?